File: packages/sdk-go/files.go
package taskforceai

import (
	"context"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
)

type multipartFormWriter interface {
	CreateFormFile(fieldname, filename string) (io.Writer, error)
	WriteField(fieldname, value string) error
	Close() error
	FormDataContentType() string
}

var newMultipartUploadWriter = func(w io.Writer) multipartFormWriter {
	return multipart.NewWriter(w)
}

type multipartUploadField struct {
	Name       string
	Value      string
	WriteError string
}

type multipartUploadBodyOptions struct {
	Filename         string
	Content          io.Reader
	CreateFileError  string
	CopyContentError string
	CloseWriterError string
	Fields           []multipartUploadField
}

func newMultipartUploadBody(opts multipartUploadBodyOptions) (*io.PipeReader, string) {
	pr, pw := io.Pipe()
	writer := newMultipartUploadWriter(pw)

	go func() {
		var closeErr error
		defer func() {
			if closeErr != nil {
				_ = pw.CloseWithError(closeErr)
				return
			}
			_ = pw.Close()
		}()

		part, err := writer.CreateFormFile("file", opts.Filename)
		if err != nil {
			closeErr = fmt.Errorf("%s: %w", opts.CreateFileError, err)
			return
		}
		if _, err := io.Copy(part, opts.Content); err != nil {
			closeErr = fmt.Errorf("%s: %w", opts.CopyContentError, err)
			return
		}
		for _, field := range opts.Fields {
			if field.Value == "" {
				continue
			}
			if err := writer.WriteField(field.Name, field.Value); err != nil {
				closeErr = fmt.Errorf("%s: %w", field.WriteError, err)
				return
			}
		}
		if err := writer.Close(); err != nil {
			closeErr = fmt.Errorf("%s: %w", opts.CloseWriterError, err)
		}
	}()

	return pr, writer.FormDataContentType()
}

// File represents an uploaded file.
type File struct {
	ID        string `json:"id"`
	Filename  string `json:"filename"`
	Purpose   string `json:"purpose"`
	Bytes     int64  `json:"bytes"`
	CreatedAt int64  `json:"created_at"` // Unix timestamp in seconds.
	MimeType  string `json:"mime_type,omitempty"`
}

// FileUploadOptions contains options for uploading a file.
type FileUploadOptions struct {
	Purpose  string `json:"purpose,omitempty"` // e.g., "assistants", "fine-tune"
	MimeType string `json:"mime_type,omitempty"`
}

// FileListResponse contains a list of files.
type FileListResponse struct {
	Files []File `json:"files"`
	Total int    `json:"total"`
}

// UploadAttachment uploads a transient task attachment and returns its attachment ID.
func (c *Client) UploadAttachment(ctx context.Context, filename string, content io.Reader, mimeType string) (string, error) {
	body, contentType := newMultipartUploadBody(multipartUploadBodyOptions{
		Filename:         filename,
		Content:          content,
		CreateFileError:  "create attachment file part",
		CopyContentError: "copy attachment content",
		CloseWriterError: "close attachment multipart writer",
		Fields: []multipartUploadField{{
			Name:       "mime_type",
			Value:      mimeType,
			WriteError: "write attachment mime_type field",
		}},
	})

	resp, err := c.sendMultipartRequest(ctx, apiRootURL(c.baseURL)+"/attachments/upload", body, contentType, "upload attachment")
	if err != nil {
		return "", err
	}
	defer func() { _ = resp.Body.Close() }()

	var result struct {
		ID string `json:"id"`
	}
	if err := decodeJSON(resp.Body, &result); err != nil {
		return "", err
	}
	if result.ID == "" {
		return "", fmt.Errorf("invalid attachment upload response: id is required")
	}
	return result.ID, nil
}

// UploadFile uploads a file to the API.
func (c *Client) UploadFile(ctx context.Context, filename string, content io.Reader, opts *FileUploadOptions) (*File, error) {
	fields := make([]multipartUploadField, 0, 2)
	if opts != nil {
		fields = append(fields,
			multipartUploadField{
				Name:       "purpose",
				Value:      opts.Purpose,
				WriteError: "write upload purpose field",
			},
			multipartUploadField{
				Name:       "mime_type",
				Value:      opts.MimeType,
				WriteError: "write upload mime_type field",
			},
		)
	}
	body, contentType := newMultipartUploadBody(multipartUploadBodyOptions{
		Filename:         filename,
		Content:          content,
		CreateFileError:  "create upload file part",
		CopyContentError: "copy upload file content",
		CloseWriterError: "close upload multipart writer",
		Fields:           fields,
	})

	resp, err := c.sendMultipartRequest(ctx, c.baseURL+"/files", body, contentType, "upload file")
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()

	return decodeValidated(resp.Body, "upload file", validateFile)
}

func (c *Client) sendMultipartRequest(ctx context.Context, url string, body *io.PipeReader, contentType, operation string) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, body)
	if err != nil {
		_ = body.CloseWithError(err)
		return nil, err
	}
	req.Header.Set("Content-Type", contentType)
	c.addRequestMetadata(ctx, req)

	resp, err := c.uploadRequest(req)
	if err != nil {
		return nil, err
	}
	if resp == nil {
		return nil, fmt.Errorf("failed to %s: response unavailable", operation)
	}
	keepBodyOpen := false
	defer func() {
		if !keepBodyOpen {
			_ = resp.Body.Close()
		}
	}()
	if c.responseHook != nil {
		c.responseHook(resp.StatusCode, resp.Header)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("failed to %s: status %d", operation, resp.StatusCode)
	}
	keepBodyOpen = true
	return resp, nil
}

func (c *Client) uploadRequest(req *http.Request) (*http.Response, error) {
	if c.uploadRequestHook != nil {
		return c.uploadRequestHook(req)
	}
	return c.httpClient.Do(req)
}

// ListFiles retrieves a list of uploaded files.
func (c *Client) ListFiles(ctx context.Context, limit, offset int) (*FileListResponse, error) {
	path := fmt.Sprintf("/files?limit=%d&offset=%d", limit, offset)
	return requestDecoded(c, ctx, http.MethodGet, path, nil, "list files", http.StatusOK, "file list", validateFileList)
}

// GetFile retrieves metadata for a specific file.
func (c *Client) GetFile(ctx context.Context, fileID string) (*File, error) {
	path := "/files/" + pathSegment(fileID)
	return requestDecoded(c, ctx, http.MethodGet, path, nil, "get file", http.StatusOK, "file", validateFile)
}

// DeleteFile deletes a file by ID.
func (c *Client) DeleteFile(ctx context.Context, fileID string) error {
	path := "/files/" + pathSegment(fileID)

	resp, err := c.requestSuccessful(ctx, "DELETE", path, nil, "delete file", 0)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()
	return nil
}

// DownloadFile downloads the content of a file.
func (c *Client) DownloadFile(ctx context.Context, fileID string) (io.ReadCloser, error) {
	path := "/files/" + pathSegment(fileID) + "/content"

	resp, err := c.requestSuccessful(ctx, "GET", path, nil, "download file", http.StatusOK)
	if err != nil {
		return nil, err
	}
	return resp.Body, nil
}

File: packages/sdk-rust/src/files.rs
use crate::client::TaskForceAI;
use crate::error::TaskForceAIError;
use crate::validation::{validate_file, validate_file_list};
use bytes::{Bytes, BytesMut};
use chrono::{DateTime, Utc};
use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
use reqwest::multipart::{Form, Part};
use serde::{Deserialize, Serialize};

#[cfg(not(test))]
const MAX_DOWNLOAD_SIZE_BYTES: usize = 50 * 1024 * 1024;
#[cfg(test)]
const MAX_DOWNLOAD_SIZE_BYTES: usize = 16;

const PATH_SEGMENT_ENCODE_SET: &AsciiSet = &CONTROLS
    .add(b' ')
    .add(b'"')
    .add(b'#')
    .add(b'$')
    .add(b'%')
    .add(b'&')
    .add(b'\'')
    .add(b'(')
    .add(b')')
    .add(b'*')
    .add(b'+')
    .add(b',')
    .add(b'/')
    .add(b':')
    .add(b';')
    .add(b'<')
    .add(b'=')
    .add(b'>')
    .add(b'?')
    .add(b'@')
    .add(b'[')
    .add(b'\\')
    .add(b']')
    .add(b'^')
    .add(b'`')
    .add(b'{')
    .add(b'|')
    .add(b'}');

fn encode_path_segment(segment: &str) -> String {
    utf8_percent_encode(segment, PATH_SEGMENT_ENCODE_SET).to_string()
}

/// Represents an uploaded file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct File {
    pub id: String,
    pub filename: String,
    pub purpose: String,
    pub bytes: i64,
    #[serde(with = "chrono::serde::ts_seconds")]
    pub created_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
}

/// Options for uploading a file.
#[derive(Debug, Clone, Default)]
pub struct FileUploadOptions {
    pub purpose: Option<String>,
    pub mime_type: Option<String>,
}

/// Response containing a list of files.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileListResponse {
    pub files: Vec<File>,
    pub total: i64,
}

#[derive(Debug, Deserialize)]
struct AttachmentUploadResponse {
    id: String,
}

impl TaskForceAI {
    /// Uploads a transient task attachment and returns its attachment ID.
    pub async fn upload_attachment(
        &self,
        filename: &str,
        content: Bytes,
        mime_type: Option<&str>,
    ) -> Result<String, TaskForceAIError> {
        let mut part = Part::bytes(content.to_vec()).file_name(filename.to_string());
        if let Some(mime_type) = mime_type {
            part = part
                .mime_str(mime_type)
                .map_err(|e| TaskForceAIError::Other(e.to_string()))?;
        }
        let form = Form::new().part("file", part);
        let url = format!("{}/attachments/upload", self.api_root_url());
        let response = self
            .with_sdk_headers(self.client.post(&url).multipart(form))
            .send()
            .await?;
        let status = response.status();
        if !status.is_success() {
            return Err(
                Self::api_error_from_response(response, "Failed to read error message").await,
            );
        }

        let uploaded: AttachmentUploadResponse = response.json().await?;
        if uploaded.id.trim().is_empty() {
            return Err(TaskForceAIError::Other(
                "Invalid attachment upload response: id is required".to_string(),
            ));
        }
        Ok(uploaded.id)
    }

    /// Uploads a file to the API.
    pub async fn upload_file(
        &self,
        filename: &str,
        content: Bytes,
        options: Option<FileUploadOptions>,
    ) -> Result<File, TaskForceAIError> {
        let mime_type = options
            .as_ref()
            .and_then(|o| o.mime_type.clone())
            .unwrap_or_else(|| "application/octet-stream".to_string());

        let mut form = Form::new().part(
            "file",
            Part::bytes(content.to_vec())
                .file_name(filename.to_string())
                .mime_str(&mime_type)
                .map_err(|e| TaskForceAIError::Other(e.to_string()))?,
        );

        if let Some(opts) = options {
            if let Some(purpose) = opts.purpose {
                form = form.text("purpose", purpose);
            }
            if let Some(mime_type) = opts.mime_type {
                form = form.text("mime_type", mime_type);
            }
        }

        let url = format!("{}/files", self.base_url);
        let response = self
            .with_sdk_headers(self.client.post(&url).multipart(form))
            .send()
            .await?;
        let status = response.status();

        if !status.is_success() {
            return Err(
                Self::api_error_from_response(response, "Failed to read error message").await,
            );
        }

        let file: File = response.json().await?;
        validate_file(&file, "upload file")?;
        Ok(file)
    }

    /// Retrieves a list of uploaded files.
    pub async fn list_files(
        &self,
        limit: i32,
        offset: i32,
    ) -> Result<FileListResponse, TaskForceAIError> {
        let path = format!("/files?limit={}&offset={}", limit, offset);
        let response: FileListResponse = self.request(reqwest::Method::GET, &path, None).await?;
        validate_file_list(&response)?;
        Ok(response)
    }

    /// Retrieves metadata for a specific file.
    pub async fn get_file(&self, file_id: &str) -> Result<File, TaskForceAIError> {
        let path = format!("/files/{}", encode_path_segment(file_id));
        let file: File = self.request(reqwest::Method::GET, &path, None).await?;
        validate_file(&file, "file")?;
        Ok(file)
    }

    /// Deletes a file by ID.
    pub async fn delete_file(&self, file_id: &str) -> Result<(), TaskForceAIError> {
        let path = format!("/files/{}", encode_path_segment(file_id));
        let _: serde_json::Value = self.request(reqwest::Method::DELETE, &path, None).await?;
        Ok(())
    }

    /// Downloads the content of a file.
    pub async fn download_file(&self, file_id: &str) -> Result<Bytes, TaskForceAIError> {
        let url = format!(
            "{}/files/{}/content",
            self.base_url,
            encode_path_segment(file_id)
        );
        let mut response = self.with_sdk_headers(self.client.get(&url)).send().await?;
        let status = response.status();

        if !status.is_success() {
            return Err(
                Self::api_error_from_response(response, "Failed to read error message").await,
            );
        }

        if let Some(content_length) = response.content_length() {
            if content_length > MAX_DOWNLOAD_SIZE_BYTES as u64 {
                return Err(TaskForceAIError::Other(format!(
                    "File too large to download safely ({} bytes > {} bytes)",
                    content_length, MAX_DOWNLOAD_SIZE_BYTES
                )));
            }
        }

        let mut body = BytesMut::new();
        while let Some(chunk) = response.chunk().await? {
            if body.len() + chunk.len() > MAX_DOWNLOAD_SIZE_BYTES {
                return Err(TaskForceAIError::Other(format!(
                    "File too large to download safely (limit: {} bytes)",
                    MAX_DOWNLOAD_SIZE_BYTES
                )));
            }
            body.extend_from_slice(&chunk);
        }

        Ok(body.freeze())
    }
}
